email/reporter.go

```diff
@@ -0,0 +1,551 @@
+	// a report has been opened but something went wrong after opening the
+	// report, in that case we want to try and continue to finish the report
+	if reportId == 0 {
+		return err
+	}
+
+	// set the report id and finish the report
+	report.ReportID = reportId
+	err = r.finishReport(report)
+	if err != nil {
+		logger.Errorf("failed to finish report, error '%v'", err)
+		return err
+	}
+
+	return nil
+}
+
+// finishReport will finish the report with NCMEC
+func (r *Reporter) finishReport(report database.NCMECReport) error {
+	// convenience variables
+	logger := r.staticLogger
+
+	// finish the report with NCMEC
+	var reportErr string
+	res, err := r.staticClient.finishReport(report.ReportID)
+	if err == nil && res.ResponseCode != ncmecStatusOK {
+		err = fmt.Errorf("unexpected response code %v when finishing report '%v'", res.ResponseCode, report.ID.Hex())
+	}
+	if err != nil {
+		reportErr = err.Error()
+		logger.Errorf("failed to finish report %v, err '%v'", report.ReportID, err)
+	}
+
+	// update the email and set the report err and reported flag
+	err = r.staticAbuseDatabase.UpdateReportNoLock(report, bson.M{
+		"$set": bson.M{
+			"filed":     err == nil,
+			"filed_at":  time.Now().UTC(),
+			"filed_err": reportErr,
+
+			"report_id": report.ReportID,
+		},
+	})
+	if err != nil {
+		logger.Errorf("failed to update report %v, err '%v'", report.ID, err)
+		return err
+	}
+	return nil
+}
+
+// openReport will open a report with NCMEC for the given email, it will
+// decorate the abuse email with the report id
+func (r *Reporter) openReport(entity database.NCMECReport) (uint64, error) {
+	// convenience variables
+	logger := r.staticLogger
+
+	// unmarshal the report
+	var report report
+	err := xml.Unmarshal([]byte(entity.Report), &report)
+	if err != nil {
+		return 0, fmt.Errorf("faild to unmarshal report, err %v", err)
+	}
+
+	// ensure the attributes are set
+	report.Xsi = "http://www.w3.org/2001/XMLSchema-instance"
+	report.NoNamespaceSchemaLocation = "https://report.cybertip.org/ispws/xsd"
+
+	// open the report
+	reportedAt := time.Now().UTC()
+	resp, err := r.staticClient.openReport(report)
+	if err == nil && resp.ResponseCode != ncmecStatusOK {
+		err = fmt.Errorf("unexpected response code %v when opening report '%v'", resp.ResponseCode, entity.ID.Hex())
+	}
+	if err != nil {
+		// update the email and set the report err
+		err = errors.Compose(err, r.staticAbuseDatabase.UpdateReportNoLock(entity, bson.M{
+			"$set": bson.M{
+				"filed_err": err.Error(),
+			},
+		}))
+		logger.Errorf("failed to open report, err '%v'", err)
+		return 0, err
+	}
+	reportId := resp.ReportId
+
+	// update the email and set the report id
+	err = r.staticAbuseDatabase.UpdateReportNoLock(entity, bson.M{
+		"$set": bson.M{
+			"filed_at":  reportedAt,
+			"report_id": reportId,
+		},
+	})
+	if err != nil {
+		logger.Errorf("failed to update report '%v', err '%v'", entity.ID.Hex(), err)
+		return reportId, nil
```

Comment:
Hm yeah so bit of a weird case here because I don't consider the database update a failure to open report, we go on and try to finish the report after this using the report id, if that succeeds everything's fine.

Reply:
I removed the return and added a comment to clarify

Reply:
This is great now. The comment explains why we're not returning the error and the `return` that's now gone no longer confuses the reader. Perfect.